F-FrontEnd/src/run_tests.py

```diff
@@ -0,0 +1,782 @@
+                                working_dir_is_tmp) for test_case in test_cases]
+                    results = [None] * num_test_cases
+                    num_failed_testcases = 0
+                    num_finished_testcases = 0
+                    test_case_num_by_name = {test_cases[i]: i for i in range(0, num_test_cases)}
+                    with multiprocessing.Pool(self.args.jobs_num) as pool:
+                        for args, res in pool.imap_unordered(run_test_case_functor, mp_args):
+                            num_finished_testcases += 1
+                            test_case = args[1]
+                            test_case_num = test_case_num_by_name[test_case]
+                            results[test_case_num] = (test_case, res)
+                            if not res.result:
+                                num_failed_testcases += 1
+                            if print_progress:
+                                print('[%s of %s] %s' % (num_finished_testcases, num_test_cases, test_case))
+                                if not res.result:
+                                    print(res.text_summary())
+                    if num_failed_testcases > 0:
+                        return RC.FAILURE, results
+            else:
+                test_case = self.args.test_case
+                if print_progress:
+                    print(test_case)
+                res = run_test_case(working_dir, test_case, self.args, testcase_deps[test_case], False,
+                                    working_dir_is_tmp)
+                results = [(test_case, res)]
+                if not res.result:
+                    if print_progress:
+                        print(res.text_summary())
+                    return RC.FAILURE, results
+        return RC.SUCCESS, results
+
+    def run(self, print_progress=False) -> TestResult:
+        start_ts = dt.datetime.utcnow()
+        return_code, results = self.__run(print_progress)
+        end_ts = dt.datetime.utcnow()
+        res = TestResult(testcases=tuple(results),
+                         result=return_code == RC.SUCCESS,
+                         start_timestamp=start_ts, end_timestamp=end_ts)
+        if print_progress:
+            print('Elapsed time: ', end_ts - start_ts)
+        if self.args.report_file is not None:
+            res.save_to_file(self.args.report_file)
+        return res
+
+    @staticmethod
+    def scan_for_dependencies(dir_path: str,
+                              debug_output: bool = False) -> Tuple[Tuple[str, ...], Dict[str, Tuple[str, ...]]]:
+        test_cases = []
+        reg_comment = re.compile('!.*')
+        spaces = '[\s]+'
+        opt_spaces = '[\s]*'
+        fortran_id = '[a-z][a-z0-9\_]*'
+        module_decl = 'module' + spaces + ('(%s)' % fortran_id)
+        submodule_name = '[\sa-z0-9\_]+'
+        submodule_decl = 'submodule' + opt_spaces + ('\((%s)\)' % submodule_name) + opt_spaces + ('(%s)' % fortran_id)
```

Comment:
Fortran is not a regular language, so cannot be parsed with regex. That being said, I am only parsing for very limited set of features related to dependencies between files. Regex are good enough for that.
That being said, it is currently unclear if this dependency scanner will work with more complicated usages. I expect that it will require further modifications, so it is more of a work in progress.
However, it worked for all the tests packed with OMNI compiler, so it is good enough for now.

Reply:
Ok, yea I think that's a very reasonable and good approach.